cirq-google/cirq_google/devices/named_topologies_test.py

```diff
@@ -0,0 +1,106 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import itertools
+from typing import Iterable
+from unittest.mock import MagicMock
+
+import cirq
+import networkx as nx
+import pytest
+from cirq_google import (
+    draw_gridlike,
+    LineTopology,
+    TiltedSquareLattice,
+    get_placements,
+    draw_placements,
+)
+import cirq_google as cg
+
+
+def test_tilted_square_lattice():
+    width = 2
+    height = 3
+    topo = TiltedSquareLattice(width, height)
+    assert all(1 <= topo.graph.degree[node] <= 4 for node in topo.graph.nodes)
+    assert topo.name == 'tilted-square-lattice-2-3'
+    assert topo.n_nodes == topo.graph.number_of_nodes()
+
+    with pytest.raises(ValueError):
+        _ = TiltedSquareLattice(0, 3)
+    with pytest.raises(ValueError):
+        _ = TiltedSquareLattice(3, 0)
+
+    ax = MagicMock()
+    topo.draw(ax=ax)
+    ax.scatter.assert_called()
+
+    qubits = topo.nodes_as_gridqubits()
+    assert all(isinstance(q, cirq.GridQubit) for q in qubits)
+
+
+def test_tilted_square_lattice_n_nodes():
+    for width, height in itertools.product(list(range(1, 4 + 1)), repeat=2):
+        topo = TiltedSquareLattice(width, height)
+        assert topo.n_nodes == topo.graph.number_of_nodes()
+
+
+def test_line_topology():
+    n = 10
+    topo = LineTopology(n)
+    assert topo.n_nodes == n
+    assert topo.n_nodes == topo.graph.number_of_nodes()
+    assert all(1 <= topo.graph.degree[node] <= 2 for node in topo.graph.nodes)
+    assert topo.name == 'line-10'
+
+    ax = MagicMock()
+    topo.draw(ax=ax)
+    ax.scatter.assert_called()
+
+    with pytest.raises(ValueError, match='greater than 1.*'):
+        _ = LineTopology(1)
+    assert LineTopology(2).n_nodes == 2
+    assert LineTopology(2).graph.number_of_nodes() == 2
+
+
+@pytest.mark.parametrize('cartesian', [True, False])
+def test_draw_gridlike(cartesian):
+    graph = nx.grid_2d_graph(3, 3)
+    ax = MagicMock()
+    pos = draw_gridlike(graph, cartesian=cartesian, ax=ax)
+    ax.scatter.assert_called()
+    for (row, column), _ in pos.items():
+        assert 0 <= row < 3
+        assert 0 <= column < 3
+
+
+def _gridqubits_to_graph_device(qubits: Iterable[cirq.GridQubit]):
+    # cirq contrib routing --> gridqubits_to_graph_device
+    def _manhattan_distance(qubit1: cirq.GridQubit, qubit2: cirq.GridQubit) -> int:
+        return abs(qubit1.row - qubit2.row) + abs(qubit1.col - qubit2.col)
+
+    return nx.Graph(
+        pair for pair in itertools.combinations(qubits, 2) if _manhattan_distance(*pair) == 1
+    )
```

Comment:
Alternative that uses existing cirq function to compute L1 norm:

```python
return nx.Graph((a, b) for (a, b) in itertools.combinations(qubits, 2) if a.is_adjacent(b))
```

Reply:
This is a copy of a cirq.contrib.routing function that I need in lots of tests I write. I was promised it would make its way to cirq-core eventually and I could replace all the places I copied it into test files. I'll just leave it for now.

In any event: this function should be implemented by searching in the (known) local neighborhood of a gridqubit rather than going over all pairs, which may be nowhere near each other.

Reply:
moving to cirq-core, I shouldn't (can't?) use Syc23 in my test, so I switched it to `TiltedSquareLattice(8, 4).graph` which is the Syc23 graph.